crypto/Pos/Crypto/Configuration.hs

```diff
@@ -1,20 +1,178 @@
+                         ", acceptable values are RequiresNoMagic | RequiresMagic")
+
+-- Canonical JSON instances
+instance Monad m => ToJSON m RequiresNetworkMagic where
+    toJSON RequiresNoMagic = pure (JSString "RequiresNoMagic")
+    toJSON RequiresMagic   = pure (JSString "RequiresMagic")
+
+instance MonadError SchemaError m => FromJSON m RequiresNetworkMagic where
+    fromJSON = \case
+        (JSString "RequiresNoMagic") -> pure RequiresNoMagic
+        (JSString "RequiresMagic")    -> pure RequiresMagic
+        other ->
+            expected "RequiresNoMagic | RequiresMagic" (Just (show other))
+
+
+--------------------------------------------------------------------------------
+-- ProtocolMagic
+--------------------------------------------------------------------------------
+
+newtype ProtocolMagicId = ProtocolMagicId
+    { unProtocolMagicId :: Int32
+    } deriving (Show, Eq, NFData)
+
+deriveSafeCopySimple 0 'base ''ProtocolMagicId
 
 -- | Magic number which should differ for different clusters. It's
 -- defined here, because it's used for signing. It also used for other
 -- things (e. g. it's part of a serialized block).
-newtype ProtocolMagic = ProtocolMagic
-    { getProtocolMagic :: Int32
-    } deriving (Show, Eq, NFData)
+--
+-- mhueschen: As part of CO-353 I am adding `getRequiresNetworkMagic` in
+-- order to pipe configuration to functions which must generate & verify
+-- Addresses (which now must be aware of `NetworkMagic`).
+data ProtocolMagic = ProtocolMagic
+    { getProtocolMagicId      :: !ProtocolMagicId
+    , getRequiresNetworkMagic :: !RequiresNetworkMagic
+    } deriving (Eq, Show, Generic)
 
-deriving instance ToJSON ProtocolMagic
-deriving instance FromJSON ProtocolMagic
+-- mhueschen: For backwards-compatibility reasons, I redefine this function
+-- in terms of the two record accessors.
+getProtocolMagic :: ProtocolMagic -> Int32
+getProtocolMagic = unProtocolMagicId . getProtocolMagicId
 
+instance NFData ProtocolMagic
 deriveSafeCopySimple 0 'base ''ProtocolMagic
+
+instance A.ToJSON ProtocolMagic where
+    toJSON (ProtocolMagic (ProtocolMagicId ident) rnm) =
+        A.object ["pm" .= ident, "requiresNetworkMagic" .= rnm]
+
+-- Here we default to `RequiresMagic` (what testnets use) if only
+-- a ProtocolMagic identifier is provided.
+instance A.FromJSON ProtocolMagic where
+    parseJSON v@(A.Number _) = ProtocolMagic
+        <$> (ProtocolMagicId <$> A.parseJSON v)
+        <*> pure RequiresMagic
+    parseJSON (A.Object o) =
+        case HM.lookup "requiresNetworkMagic" o of
+            -- legacy codebase JSON encoding
+            Just (A.String "NMMustBeJust")
+                -> ProtocolMagic <$> (ProtocolMagicId <$> o .: "pm")
+                                 <*> pure RequiresMagic
+            Just (A.String "NMMustBeNothing")
+                -> ProtocolMagic <$> (ProtocolMagicId <$> o .: "pm")
+                                 <*> pure RequiresNoMagic
```

Comment:
Is there a reason that this legacy-migration logic is in the `FromJSON ProtocolMagic` instance, rather than directly in the `FromJSON RequiresNetworkMagic` instance?

Do we have an issue [here](https://github.com/input-output-hk/cardano-sl/pull/3715/files#diff-948e2bae1555211ce53399a3df371c30R98) and [here](https://github.com/input-output-hk/cardano-sl/pull/3715/files#diff-948e2bae1555211ce53399a3df371c30R104) with supporting the legacy names? It seems like those use the `FromJSON RequiresNetworkMagic` instance, which will not be equipped to handle the legacy names.

Reply:
You're absolutely right. I'll make the necessary changes.